### PR TITLE
[RFC] options: set 'laststatus' to 2 by default

### PR DESCRIPTION
**Pull request by fmoralesc**
Re: https://github.com/neovim/neovim/issues/2676

### CODE REVIEW

**Comment by fmoralesc** [comment]
This also impacted a whole lot of tests. Since `ls=2` can clutter the tests which never use more than one window, I decided to use `ls=1` as a general setting, but also introduced a test to check that dthe default was really applied (like with `listchars`, I needed to apply the option on startup, it didn't catch otherwise).

**Comment by justinmk** [comment]
I'm leaning towards thinking this change is not worth it. `laststatus=1` has merits if you don't use a fancy statusline (`set ruler` works without the statusline, and it saves a line if you have only 1 window open.

What do you think?

**Comment by fmoralesc** [comment]
I would agree, but since we aren't introducing 'ruler', I'm leaning towards `laststatus=2` as a better default.

**Comment by fmoralesc** [comment]
At any rate, the change in https://github.com/fmoralesc/neovim/commit/72d1624148ba6f3d90dd6e92584e007a066b4507 should be merged anyway, because that fixes a proper bug in the implementation of laststatus (vim just never caught it because of the old default).

File: test/functional/helpers.lua [hunk 1]
```diff
@@ -6,9 +6,10 @@ local AsyncSession = require('nvim.async_session')
 local Session = require('nvim.session')
 
 local nvim_prog = os.getenv('NVIM_PROG') or 'build/bin/nvim'
---- FIXME: 'autoindent' messes up the insert() function
+-- FIXME: 'autoindent' messes up the insert() function
+-- 'ls=1' makes the screen related tests less cluttered
 local nvim_argv = {nvim_prog, '-u', 'NONE', '-i', 'NONE', '-N',
-                   '--cmd', 'set shortmess+=I background=light noswapfile noautoindent',
+                   '--cmd', 'set shortmess+=I background=light noswapfile noautoindent ls=1',
```
Comment (justinmk) [inline test/functional/helpers.lua]: use full option name

File: test/functional/ui/screen_basic_spec.lua [hunk 2]
```diff
@@ -79,6 +87,31 @@ describe('Screen', function()
...
+        ~                                                    |
+        ~                                                    |
+        ~                                                    |
+        [No Name]                                            |
+                                                             |
+      ]])
+    end)
+  end)
+
+  describe('with ls=1', function()
```
Comment (justinmk) [inline test/functional/ui/screen_basic_spec.lua]: use full name

File: test/functional/ui/screen_basic_spec.lua [hunk 3]
```diff
@@ -1,13 +1,21 @@
...
+                   '--cmd', 'set shortmess+=I background=light noswapfile noautoindent',
+                   '--embed'}
 
   before_each(function()
-    clear()
+    if session then
+      session:exit(0)
+    end
+    local screen_nvim = spawn(nvim_argv)
+    set_session(screen_nvim)
```
Comment (justinmk) [inline test/functional/ui/screen_basic_spec.lua]: I'm confused about why this as needed: it initializes a session with laststatus=2, yet none of the tests were changed. Is it still needed?

File: test/functional/ui/screen_basic_spec.lua [hunk 4]
```diff
@@ -79,6 +87,31 @@ describe('Screen', function()
     end)
   end)
 
+  describe('startup', function()
+    it('default startup screen', function()
```
Comment (justinmk) [inline test/functional/ui/screen_basic_spec.lua]: IIUC this is the only test that needs laststatus=2. I think it would be clearer if the `set_session` were inlined inside this single `it()`. Or relocated to a separate describe() group.

Comment (fmoralesc) [inline test/functional/ui/screen_basic_spec.lua]: Yes, I should do that. This is a test for the behavior the 2nd commit should fix, it makes sure the laststatus value is applied.

**Comment by fmoralesc** [comment]
Made the changes @justinmk proposed, and rebased.

**Comment by fmoralesc** [comment]
Rebased.

**Comment by fmoralesc** [comment]
Did the clint rules change, @justinmk? I'm getting a bunch of errors for stuff not directly related to this PR, and I feel it might cascade.

**Comment by justinmk** [comment]
Yes, all *changes* to all files are checked by clint now thanks to @ZyX-I.

**Comment by fmoralesc** [comment]
OK. Meanwhile, I fixed the issues more directly related to the PR, but perhaps I should go ahead and change `last_status()` so it takes a `bool` instead of an int, and fix the issues related to that too? I'm not sure how deep that rabbit hole goes, though.

File: test/functional/ui/screen_basic_spec.lua (hunk 3 above)
Comment (fmoralesc) [inline test/functional/ui/screen_basic_spec.lua]: BTW, this was done like this so the original tests didn't need to be changed (they assume laststatus=1).

File: test/functional/helpers.lua [hunk 5]
```diff
@@ -6,8 +6,10 @@ local AsyncSession = require('nvim.async_session')
 local Session = require('nvim.session')
 
 local nvim_prog = os.getenv('NVIM_PROG') or 'build/bin/nvim'
+-- FIXME: 'autoindent' messes up the insert() function
+-- 'laststatus=1' makes the screen related tests less cluttered
```
Comment (justinmk) [inline test/functional/helpers.lua]: It is understood that these `--cmd` settings are meant to work around and/or enforce the expectations of the tests, so I removed these comments.

File: test/functional/ui/screen_basic_spec.lua [hunk 6]
```diff
@@ -1,8 +1,51 @@
...
+local spawn, set_session, clear = helpers.spawn, helpers.set_session, helpers.clear
+local feed, execute = helpers.feed, helpers.execute
 local insert, wait = helpers.insert, helpers.wait
 
+describe('Initial screen', function()
+  local screen
+  local nvim_prog = os.getenv('NVIM_PROG') or 'build/bin/nvim'
+  local nvim_argv = {nvim_prog, '-u', 'NONE', '-i', 'NONE', '-N',
+	'--cmd', 'set shortmess+=I background=light noswapfile',
+	'--embed'}
```
Comment (justinmk) [inline test/functional/ui/screen_basic_spec.lua]: spaces not tabs

File: test/functional/ui/screen_basic_spec.lua [hunk 7]
```diff
@@ -1,8 +1,51 @@
 local helpers = require('test.functional.helpers')
 local Screen = require('test.functional.ui.screen')
-local clear, feed, execute = helpers.clear, helpers.feed, helpers.execute
+local spawn, set_session, clear = helpers.spawn, helpers.set_session, helpers.clear
+local feed, execute = helpers.feed, helpers.execute
 local insert, wait = helpers.insert, helpers.wait
 
+describe('Initial screen', function()
+  local screen
+  local nvim_prog = os.getenv('NVIM_PROG') or 'build/bin/nvim'
```
Comment (justinmk) [inline test/functional/ui/screen_basic_spec.lua]: why not use `helpers.nvim_prog`?

File: src/nvim/option.c [hunk 8]
```diff
@@ -454,6 +454,9 @@ void set_init_1(void)
   /* Set all options to their Vim default */
   set_options_default(OPT_FREE);
 
+  // set 'laststatus'
+  last_status(false);
```
Comment (justinmk) [inline src/nvim/option.c]: why is this needed? why isn't `set_options_default()` enough?

Comment (bfredl) [inline src/nvim/option.c]: `set_options_default()` is not enough. One sees this by (on master) `set ls=2` and then `set all&` (which calls `set_options_default()`) . Now `set ls` shows 1 but nvim still behaves _as if_ `&ls=2`.

Comment (ZyX-I) [inline src/nvim/option.c]: @bfredl I think that this is clearly a `set_options_default` bug which should be fixed in `set_options_default` body and not worked around here.

Comment (bfredl) [inline src/nvim/option.c]: :+1:

Comment (justinmk) [inline src/nvim/option.c]: Yes, `set_num_option` seems to handle it, so I wonder why that isn't happening from `set_options_default`.

File: test/functional/ui/screen_basic_spec.lua (hunk 7 above)
Comment (fmoralesc) [inline test/functional/ui/screen_basic_spec.lua]: Didn't see it at the time, thanks for the hint ;)

**Comment by justinmk** [comment]
Merged. Thanks @fmoralesc
